### ISSUE TITLE
GRR not taking Clients.server_url nor EXTERNAL_HOSTNAME in account

### ISSUE THREAD

**Issue by sarcarx** [google/grr#392]
I am using version v0.3.0-7 of the docker image.
I have set EXTERNAL_HOSTNAME to "grr_server"
I downloaded the client from the web UI and started it in a virtual machine on Windows 7.
Process Monitor shows the GRR process tring to contact localhost:8080. Starting the debbugged GRR.exe process with `GRR.exe --config=GRR.exe.yaml --verbose` shows the following error:
```
INFO:2016-06-09 18:10:17,269 logging:1622] Could not connect to GRR servers ['ht
tp://localhost:8080/control'], directly or through these proxies: [''].
```
I also tried editing the GRR.exe.yaml file directly, setting Clients.server_urls to another url. The error is the same.
I tried with a name, and with an IP, I got the same result.
I believe the grr.exe is supposed to take in account the modifications of the yaml file, which isn't the case. What am I missing?

**Comment by sarcarx** [comment]
It seems `Client.control_urls` works instead of `Client.server_urls`. 
I changed the yaml file to:
`Client.control_urls: http://grr_server:8080`
so now the error is:
`INFO:2016-06-09 18:21:52,467 logging:1622] Could not connect to GRR servers ['ht
tp://grr_server:8080'], directly or through these proxies: [''].`
I tried to ping the server and it worked. Process Monitor and wireshark show no sign of paquets sent to this IP.

**Comment by grrrrrrrrr** [comment]
Can you d/l the server.pem? Is the GRR server running there?

Have a look at
https://github.com/google/grr-doc/blob/master/troubleshooting.adoc#i-dont-see-my-clients

**Comment by sarcarx** [comment]
Yes the server.pem file is perfectly downloadable.
But the error I have is a bit different than the one I see in the doc, mine seems like it has detected a proxy with an empty name.

**Comment by sarcarx** [comment]
The server.pem is downloadable.
I just noticed another error, my bad:
`INFO:2016-06-09 18:34:04,280 logging:1622] Unable to verify server certificate a
t http://grr_server:8080/server.pem: Server cert is invalid.`
It appeared after I added the right Clients.control_urls parameter

**Comment by grrrrrrrrr** [comment]
The empty proxy name means no proxy used and is always present. Do you need
to use a proxy? Can you maybe post the client output again with the url
fixed?

**Comment by sarcarx** [comment]
Let me give you the entire client output so I don't waste your time on my mistakes:
`INFO:2016-06-09 18:34:03,168 logging:1622] Starting GRR Prelogging buffer.
INFO:2016-06-09 18:34:03,219 logging:1622] Loading configuration from GRR.exe.ya
ml
WARNING:2016-06-09 18:34:03,219 logging:1612] Missing config definition for Clie
nt.server_urls. This option is likely deprecated or renamed. Check the release n
otes.
WARNING:2016-06-09 18:34:03,219 logging:1612] Missing config definition for Clie
nt.build_environment. This option is likely deprecated or renamed. Check the rel
ease notes.
INFO:2016-06-09 18:34:03,219 logging:1622] Loading configuration from reg://HKEY
_LOCAL_MACHINE/Software\GRR
INFO:2016-06-09 18:34:03,219 logging:1622] Configuration writeback is set to reg
://HKEY_LOCAL_MACHINE/Software\GRR
DEBUG:2016-06-09 18:34:03,219 logging:1630] Initializing Logging subsystem.
DEBUG:2016-06-09 18:34:03,219 logging:1630] Will use logging engines ['file', 's
tderr', 'event_log']
DEBUG:2016-06-09 18:34:03,219 logging:1630] Applying filter env for WINDIR.
INFO:2016-06-09 18:34:03,219 logging:1622] Writing log file to C:\Windows\System
32\logfiles\\GRR.log
DEBUG:2016-06-09 18:34:03,229 logging:1630] Initializing StatsInit, required by
CommunicatorInit
DEBUG:2016-06-09 18:34:03,229 logging:1630] Initializing CommunicatorInit
DEBUG:2016-06-09 18:34:03,229 logging:1630] Initializing VFSInit
DEBUG:2016-06-09 18:34:03,229 logging:1630] Initializing CommsInit
DEBUG:2016-06-09 18:34:03,229 logging:1630] Initializing InitHook
DEBUG:2016-06-09 18:34:03,229 logging:1630] Initializing PluginLoader
DEBUG:2016-06-09 18:34:03,229 logging:1630] Applying filter env for SystemRoot.
DEBUG:2016-06-09 18:34:03,239 logging:1630] Applying filter env for SystemRoot.
DEBUG:2016-06-09 18:34:03,239 logging:1630] Applying filter env for SystemRoot.
DEBUG:2016-06-09 18:34:03,239 logging:1630] Applying filter env for WINDIR.
INFO:2016-06-09 18:34:04,270 logging:1622] Starting client aff4:/C.56a3761260690
081
DEBUG:2016-06-09 18:34:04,280 logging:1630] Sending startup information.
DEBUG:2016-06-09 18:34:04,280 logging:1630] Found proxy servers: []
INFO:2016-06-09 18:34:04,280 logging:1622] Unable to verify server certificate a
t http://grr_server:8080/server.pem: Server cert is invalid.
INFO:2016-06-09 18:34:04,280 logging:1622] Could not connect to GRR servers ['ht
tp://grr_server:8080/'], directly or through these proxies: [''].`

**Comment by grrrrrrrrr** [comment]
Thanks! It looks like your client downloaded the certificate but doesn't
accept it as valid. This often means that there is a captive portal or
something in between that gives you something else instead of the real cert
but in your case, I suspect that the client simply has the wrong
configuration. The config contains keys and certs to verify the server and
if the control urls have not been copied, chances are that you also don't
have the proper keys that were generated for you.

I'll try to set up a docker image myself to see if this is working for me
or why the config doesn't get copied properly.

**Comment by sarcarx** [comment]
the server cert is invalid because I used an old binary, let me try with a new one and i will come back to you.

**Comment by sarcarx** [comment]
It works!
Ok so it seemed that the `Client.server_urls` parameter should be replaced by `Client.control_urls` in the self-extracting binary, I understand I have to rebuild the clients. I do have a question though, why isn't that automatic when I assigned `EXTERNAL_HOSTNAME` to the right domain name?
Thanks a lot for your help.

**Comment by grrrrrrrrr** [comment]
Does the server_urls entry point at the correct location? server_urls is
the new name for this parameter so I assume that the docker image now only
works with the new server version, not with 0.3.0.7 anymore. I don't think
we will fix this anymore, the new server is basically done at this point
and we will rather focus on getting it out and make the docker image use
that one by default instead. 0.3.0.7 is pretty aged at this point...

Thanks for reporting this!

**Comment by sarcarx** [comment]
ok, I didn't use the latest version of the docker image because I had an error while running it. Should I open another issue or use this one?

**Comment by destijl** [comment]
If the lastest docker build doesn't work in some way please open a new bug with the details. I suspect the control_urls problem is the same as this one:
https://github.com/google/grr/issues/387

which I'm debugging now. I'll leave this open until I can confirm.

**Comment by destijl** [comment]
I think this should be fixed with the fix to #387
